### PR TITLE
Group Tests

### PR DESCRIPTION
**Pull request by dANW34V3R**
Add infrastructure for testing the groups assigned to each micro-op of a macro-op, along with 3 example tests but no extensive usage. Expected micro-op groups should be provided in the order which micro-ops are held in the `MacroOp` structure.

Code in `RegressionTest.cc` has been refactored to reduce repetition. This is a little hard to decipher purely from the diff provided by GitHub. Similar reductions in repetition have also been performed in the child classes.

I plan to provide more extensive testing in a future PR.

### CODE REVIEW

File: test/regression/riscv/RISCVRegressionTest.cc [hunk 1]
```diff
@@ -3,15 +3,24 @@
 #include "simeng/arch/riscv/Architecture.hh"
 #include "simeng/pipeline/BalancedPortAllocator.hh"
 
+using MacroOp = std::vector<std::shared_ptr<simeng::Instruction>>;
```
Comment (dANW34V3R) [inline test/regression/riscv/RISCVRegressionTest.cc]: Now unused. Remove

File: test/regression/aarch64/MicroOperation.cc [hunk 2]
```diff
@@ -10,6 +10,7 @@
 namespace {
 
 using MicroOp = AArch64RegressionTest;
+using namespace simeng::arch::aarch64;
```
Comment (dANW34V3R) [inline test/regression/aarch64/MicroOperation.cc:13]: Could add `InstructionGroups` to namespace so this doesn't need to be repeatedly qualified in the tests. Thoughts @UoB-HPC/simeng-developers ?

File: test/regression/aarch64/AArch64RegressionTest.cc [hunk 3]
```diff
@@ -3,22 +3,23 @@
 #include "simeng/arch/aarch64/Architecture.hh"
 #include "simeng/pipeline/BalancedPortAllocator.hh"
 
+using MacroOp = std::vector<std::shared_ptr<simeng::Instruction>>;
```
Comment (dANW34V3R) [inline test/regression/aarch64/AArch64RegressionTest.cc]: No unused, remove

File: test/regression/aarch64/MicroOperation.cc (hunk 2 above)
Comment (dANW34V3R) [reply test/regression/aarch64/MicroOperation.cc:13]: Doing this makes development much easier so done in #420

File: test/regression/RegressionTest.cc [hunk 4]
```diff
@@ -52,51 +50,71 @@ void RegressionTest::run(const char* source, const char* triple,
...
+  architecture_ = instantiateArchitecture(*kernel_);
+}
 
-  // Create a port allocator for an out-of-order core
-  std::unique_ptr<simeng::pipeline::PortAllocator> portAllocator =
-      createPortAllocator();
+void RegressionTest::instantiateMemoryInterfaces() {
+  // Create memory interfaces for instruction and data access.
+  // For each memory interface, a dereferenced shared_ptr to the
+  // processImage is passed as argument.
```
Comment (ABenC377) [inline test/regression/RegressionTest.cc]: argument is missing an article ('an' or 'the').

File: test/regression/RegressionTest.cc [hunk 5]
```diff
@@ -52,51 +50,71 @@ void RegressionTest::run(const char* source, const char* triple,
...
   // Create the architecture
-  architecture_ = createArchitecture(kernel);
+  architecture_ = instantiateArchitecture(*kernel_);
+}
 
-  // Create a port allocator for an out-of-order core
-  std::unique_ptr<simeng::pipeline::PortAllocator> portAllocator =
-      createPortAllocator();
+void RegressionTest::instantiateMemoryInterfaces() {
+  // Create memory interfaces for instruction and data access.
```
Comment (ABenC377) [inline test/regression/RegressionTest.cc:95]: There aren't full stops at the end of other comments in this file.  I think we should be consistent one way or the other.

File: test/regression/RegressionTest.hh [hunk 6]
```diff
@@ -71,19 +71,29 @@ class RegressionTest
   /** Generate a default YAML-formatted configuration. */
   virtual void generateConfig() const = 0;
 
+  /** Instantiate the architecture object using the kernel */
```
Comment (ABenC377) [inline test/regression/RegressionTest.hh]: I think its confusing to say that the createArchitecture function Instantiates the architecture, when there is another function later on called instantiateArchitecture.  Could you Change comment to 'Create the ...'

Also, the other comments in this file have a full stop

File: test/regression/RegressionTest.hh [hunk 7]
```diff
@@ -71,19 +71,29 @@ class RegressionTest
...
+  void createArchitecture(const char* source, const char* triple,
+                          const char* extensions);
+
   /** Run the assembly in `source`, building it for the target `triple` and ISA
    * extensions. */
   void run(const char* source, const char* triple, const char* extensions);
 
   /** Create an ISA instance from a kernel. */
```
Comment (ABenC377) [inline test/regression/RegressionTest.hh]: As above, I'd change the verb here to 'Instantiate'

File: test/regression/RegressionTest.hh [hunk 8]
```diff
@@ -71,19 +71,29 @@ class RegressionTest
...
-  virtual std::unique_ptr<simeng::arch::Architecture> createArchitecture(
+  virtual std::unique_ptr<simeng::arch::Architecture> instantiateArchitecture(
       simeng::kernel::Linux& kernel) const = 0;
 
   /** Create a port allocator for an out-of-order core model. */
   virtual std::unique_ptr<simeng::pipeline::PortAllocator> createPortAllocator(
       ryml::ConstNodeRef config =
           simeng::config::SimInfo::getConfig()) const = 0;
 
+  /** Predecode the first instruction in source and check the assigned group
```
Comment (ABenC377) [inline test/regression/RegressionTest.hh:92]: full stop

File: test/regression/RegressionTest.hh [hunk 9]
```diff
@@ -108,41 +118,81 @@ class RegressionTest
...
-  /** The process memory. */
-  char* processMemory_ = nullptr;
+ private:
+  /** Assemble test source to a flat binary for the given triple and ISA
+   * extensions. */
+  void assemble(const char* source, const char* triple, const char* extensions);
 
-  /** The size of the process memory in bytes. */
-  size_t processMemorySize_ = 0;
+  /** Instantiate the process from the source bytes for the given architecture.
```
Comment (ABenC377) [inline test/regression/RegressionTest.hh]: As we are using instantiate and create to mean different things previously, can you change comment to use Create

File: test/regression/RegressionTest.hh [hunk 10]
```diff
@@ -108,41 +118,81 @@ class RegressionTest
...
+   */
+  void createProcess(const char* source, const char* triple,
+                     const char* extensions);
 
-  /** The process that was executed. */
-  std::unique_ptr<simeng::kernel::LinuxProcess> process_;
+  /** Instantiate the kernel object using the process. */
```
Comment (ABenC377) [inline test/regression/RegressionTest.hh]: As above, change comment to use Create

File: test/regression/riscv/RISCVRegressionTest.hh [hunk 11]
```diff
@@ -53,49 +53,102 @@ inline std::string paramToString(
...
 
-/** The test fixture for all RISCV regression tests. */
+/** A helper macro to predecode the first instruction in a snippet of RISC-V
+ * assembly code and check the assigned group(s) for each micro-op matches the
+ * expected group(s). Returns from the calling function if a fatal error occurs.
+ * Four bytes containing zeros are appended to the source to ensure that the
+ * program will terminate with an unallocated instruction encoding exception
+ * instead of running into the heap.
+ */
+#define EXPECT_GROUP(source, ...)                                   \
```
Comment (ABenC377) [inline test/regression/riscv/RISCVRegressionTest.hh:91]: Is this the same as the macro defined in the aarch64RegressionTest.hh file?  If so, could it be moved to the generic RegressionTest.hh file to avoid duplication?

File: test/regression/riscv/RISCVRegressionTest.hh [hunk 12]
```diff
@@ -53,49 +53,102 @@ inline std::string paramToString(
...
+
+  /** Run the first instruction in source through predecode and check the
+   * groups. */
+  void checkGroup(const char* source, const std::vector<int> expectedGroups,
+                  bool compressed);
 
   /** Generate a default YAML-formatted configuration. */
   void generateConfig() const override;
 
   /** Create an ISA instance from a kernel. */
```
Comment (ABenC377) [inline test/regression/riscv/RISCVRegressionTest.hh]: Again, create versus instantiate

**Review by ABenC377 — CHANGES_REQUESTED** [review]
Looks good!  Mainly minor comments about the use of the words Create and Instantiate throughout.

File: test/regression/RegressionTest.cc (hunk 5 above)
Comment (dANW34V3R) [reply test/regression/RegressionTest.cc:95]: There are a few, mostly from old comments. Unsure of the best course of action as we are very inconsistent throughout the project. @jj16791 @FinnWilkinson

File: test/regression/riscv/RISCVRegressionTest.hh (hunk 11 above)
Comment (dANW34V3R) [reply test/regression/riscv/RISCVRegressionTest.hh:91]: These are not the same. They call their respective checkGroup function which is ISA specific. Both initialise LLVM for their respective ISA and the RISC-V version takes an extra parameter for compressed instructions

File: test/regression/RegressionTest.hh (hunk 6 above)
Comment (dANW34V3R) [reply test/regression/RegressionTest.hh]: Updated

File: test/regression/RegressionTest.cc (hunk 5 above)
Comment (ABenC377) [reply test/regression/RegressionTest.cc:95]: It's not all that important.  Would be nice to have a house preference though

**Review by ABenC377 — DISMISSED** [review]
👌🏻

File: test/regression/RegressionTest.cc [hunk 13]
```diff
@@ -52,51 +50,71 @@ void RegressionTest::run(const char* source, const char* triple,
...
   // member processMemory_.
   std::shared_ptr<char> procImgPtr = process_->getProcessImage();
   processMemory_ = procImgPtr.get();
 
-  // Create memory interfaces for instruction and data access.
-  // For each memory interface, a dereferenced shared_ptr to the
-  // processImage is passed as argument.
-  simeng::memory::FlatMemoryInterface instructionMemory(processMemory_,
-                                                        processMemorySize_);
+  // TODO ensure performing this before kernel_.createProcess is ok - Think it's
```
Comment (jj16791) [inline test/regression/RegressionTest.cc]: This is valid

File: test/regression/RegressionTest.cc [hunk 14]
```diff
@@ -21,10 +21,8 @@ void RegressionTest::TearDown() {
   }
 }
 
-void RegressionTest::run(const char* source, const char* triple,
-                         const char* extensions) {
-  testing::internal::CaptureStdout();
-
+void RegressionTest::createProcess(const char* source, const char* triple,
```
Comment (jj16791) [inline test/regression/RegressionTest.cc]: What was the motivation for splitting the creation of simulation into multiple functions? It all seems a bit messy and jumbled to me

File: test/regression/RegressionTest.hh [hunk 15]
```diff
@@ -108,41 +119,82 @@ class RegressionTest
...
+                    const char* extensions);
+
+  /** Instantiate the memory interfaces. */
+  void instantiateMemoryInterfaces();
+
+  /** Instantiating all of the objects used to setup and run the simulation. */
+  void instantiateSimulationObjects(const char* source, const char* triple,
+                                    const char* extensions);
+
+  /* Instantiation of the kernel. */
```
Comment (jj16791) [inline test/regression/RegressionTest.hh]: "Pointer to the instantiated ...". Same should apply to the below

File: test/regression/RegressionTest.hh [hunk 16]
```diff
@@ -108,41 +119,82 @@ class RegressionTest
   /** The initial data to populate the heap with. */
   std::vector<uint8_t> initialHeapData_;
 
-  /** The maximum number of ticks to run before aborting the test. */
-  uint64_t maxTicks_ = UINT64_MAX;
+  /** The process that was executed. */
```
Comment (jj16791) [inline test/regression/RegressionTest.hh]: The process to be executed

**Review by jj16791 — CHANGES_REQUESTED** [review]

File: test/regression/RegressionTest.cc (hunk 5 above)
Comment (jj16791) [reply test/regression/RegressionTest.cc:95]: Probably needs a formal discussion but I'd say full stop for only the DOxygen style comments

File: test/regression/RegressionTest.cc (hunk 14 above)
Comment (dANW34V3R) [reply test/regression/RegressionTest.cc]: To check what group an instruction is assigned `architecture->predecode` is called. "RegressionTest::checkGroup::184". To do this the architecture needs to be instantiated.

Previously I just had a new function with all the code up to the architecture instantiation repeated from the old `run` function. The current code is an attempt to reduce the repetition with the side effect of also breaking up the old massive function. 

I think the chained calling is a little odd but the hope was that you wouldn't have to orchestrate the instantiations yourself as they need to happen in a specific order e.g. you need `kernel_` before you can instantiate the architecture.

Looking again it does seem a little messy and not consistent. I can move the BP and core instantiations into their own functions as well which helps a little (should have just pushed). Other thoughts on cleanup?

tldr, both `run` and `checkGroup` need to instantiate some simulation objects and this reduces repeated code

**Review by jj16791 — DISMISSED** [review]

File: test/regression/RegressionTest.cc (hunk 14 above)
Comment (jj16791) [reply test/regression/RegressionTest.cc]: After offline discussion, changes made look a lot better

File: test/regression/RegressionTest.cc [hunk 17]
```diff
@@ -143,6 +156,24 @@ void RegressionTest::run(const char* source, const char* triple,
   programFinished_ = true;
 }
 
+void RegressionTest::checkGroup(const char* source, const char* triple,
+                                const char* extensions,
+                                const std::vector<int> expectedGroups) {
```
Comment (FinnWilkinson) [inline test/regression/RegressionTest.cc]: In both AArch64 and RV Instruction groups are defined as `uint16_t` so the vector should reflect this too

File: test/regression/riscv/instructions/atomic.cc
Comment (FinnWilkinson) [inline test/regression/riscv/instructions/atomic.cc:1]: These all seem to have the group of only `LOAD_INT`. Are there no atomic instructions which perform a LOAD and STORE? i.e. a swap or a load-action-store?

**Review by FinnWilkinson — CHANGES_REQUESTED** [review]
Looks pretty good, just a few minor points

**Review by JosephMoore25 — DISMISSED** [review]
Clean PR and implementation. Makes it easy to add group tests, and the background logic all looks sound. Nice work

File: test/regression/riscv/instructions/atomic.cc
Comment (dANW34V3R) [reply test/regression/riscv/instructions/atomic.cc:1]: Comment added explaining current functionality but more discussion needed about how to handle this long term

**Review by JosephMoore25 — APPROVED** [review]

**Review by FinnWilkinson — APPROVED** [review]
Looks good to me!

**Review by jj16791 — APPROVED** [review]

**Review by ABenC377 — APPROVED** [review]

**Review by jj16791 — APPROVED** [review]
